OAuth1a authorization feature implementation

acoumb opened the pull request:
Current PR contains the implementation of the `OAuth1a` feature. With this integration, a shift of paradigm is introduced from the initial authorization model established with `OAuth2`.

An overview of the two authorization models.

## OAuth2
For `OAuth2`, getting the user authenticated is a two-step process:
- Step 1:
    - Clients sends request to get authenticated and is redirected to the service's server for providing authorization credentials.
    - On successful operation, the server returns an authorization code.
- Step 2:
    - Using the authorization code from step 1, Client sends it to the service's server in exchange for an access token and a refresh token.

### Handling requests
Requests made to the API will contain the access token from the database in the authorization header. 

## OAuth1a
`OAuth1a` introduces an additional step, as the authorization flow will look like this:
- Step 1:
    a. Client sends a `POST` request to get the OAuth request token with the following parameters:
        - `oauth_callback`
        - `oauth_consumer_key` (represented by `ClientId`)
        - `oauth_nonce`
        - `oauth_signature` - `HMAC-SHA1` encrypted string of this format: `[HTTP Method(Uppercase)]&[Request URL]&[Authorization Parameters]`. The encryption key is a string with this format: `[Consumer Key (represented by ClientSecret)]&[OAuth Token]` (in this step, we do not have a token, so I pass an empty string). 
        - `oauth_signature_method=HMAC-SHA1`
        - `oauth_timestamp`
        - `oauth_version=1.0`
        Parameters need to be passed in an alphabetical order.
    b. Server's response will contain several parameters passed as query string values: `?oauth_token=...&oauth_verifier`. The OAuth token is now cached - the reason behind this is that in the `OAuth1a` flow I cannot use a state parameter to identify in the next step which service was authorized (I can pass it, but the service does not send it back), so I use the token to identify the service.
    c. Using the response previously received, the authorization URL is built.
- Step 2:
    - Clients makes an authorization request using the URL created in the previous step and is redirected to the service's server where he will provide the authorization credentials.
    - Server responds with an OAuth Verifier code.
- Step 3:
    - Using the OAuth Verifier code from Step 2, and the OAuth request token from Step 1, Client sends them in exchange for an access token.
    - Received access token is saved into the database.

### Handling requests
Requests made to the API will contain the parameters passed in Step 1.a, but additionally will include the `oauth_token`. This means that the signature will be hashed using a different key format.

In the implementation process the following services have been tested and used as reference:
- [Twitter](https://developer.twitter.com/en/docs/tutorials/authenticating-with-twitter-api-for-enterprise/oauth1-0a-and-user-access-tokens)
- [Flickr](https://www.flickr.com/services/api/auth.oauth.html)

AndyButland commented on tests/Umbraco.AuthorizedServices.Tests/Services/AuthorizedServiceAuthorizerTests.cs:
```diff
@@ -14,7 +14,8 @@ internal class AuthorizedServiceAuthorizerTests : AuthorizedServiceTestsBase
     public void SetUp()
     {
-        TokenStorageMock = new Mock<ITokenStorage>();
+        TokenStorageMock = new Mock<ITokenStorage<Token>>();
```
As well as updating the existing tests so they still pass as you have here, don't forget to also create unit tests for the new methods that you have introduced.

acoumb replied:
Revisited existing tests and corrected. Additional test cases will follow with a different commit.

AndyButland reviewed (commented):
Looking good.  I've added quite a few comments, but most of them are of the same theme - thinking we should align existing method names with the new ones now we have two different versions of OAuth supported.

There's one comment in there about the flow itself in `AuthorizedServiceCaller.cs` - that either doesn't seem quite right or I've not understood.

And I'd like to keep the test coverage up, but introducing tests for any new methods you've added.

acoumb commented on src/Umbraco.AuthorizedServices/Controllers/AuthorizedServiceController.cs:
```diff
@@ -198,4 +223,6 @@ public async Task<IActionResult> GenerateToken(GenerateToken model)
 
     private bool StoredTokenExists(ServiceDetail serviceDetail) => _tokenStorage.GetToken(serviceDetail.Alias) != null;
+
+    private bool StoredOAuth1aTokenExists(ServiceDetail serviceDetail) => _oauth1aTokenStorage.GetToken(serviceDetail.Alias) != null;
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Controllers/AuthorizedServiceController.cs:
```diff
@@ -19,8 +23,11 @@ namespace Umbraco.AuthorizedServices.Controllers;
 {
     private readonly IOptionsMonitor<ServiceDetail> _serviceDetailOptions;
-    private readonly ITokenStorage _tokenStorage;
+    private readonly ITokenStorage<Token> _tokenStorage;
```
Introduced `IOAuth1TokenStorage` and `IOAuth2TokenStorage` which inherit `ITokenStorage<T>`, and corrected all references.

acoumb commented on src/Umbraco.AuthorizedServices/Controllers/AuthorizedServiceResponseController.cs:
```diff
@@ -72,6 +75,27 @@ public async Task<IActionResult> HandleIdentityResponse(string code, string stat
+    /// <param name="oauth_verifier">The oauth_verifier.</param>
+    /// <returns></returns>
+    /// <exception cref="AuthorizedServiceException"></exception>
+    public async Task<IActionResult> HandleOAuth1IdentityResponse(string oauth_token, string oauth_verifier)
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Extensions/HttpContextExtensions.cs:
```diff
@@ -7,4 +7,7 @@ internal static class HttpContextExtensions
     public static string GetAuthorizedServiceRedirectUri(this HttpContext httpContext)
         => $"{httpContext.Request.Scheme}://{httpContext.Request.Host}/umbraco/api/{nameof(AuthorizedServiceResponseController).Replace("Controller", string.Empty)}/{nameof(AuthorizedServiceResponseController.HandleIdentityResponse)}";
+
+    public static string GetOAuth1AuthorizedServiceRedirectUri(this HttpContext httpContext)
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Extensions/HttpContextExtensions.cs:
```diff
@@ -7,4 +7,7 @@ internal static class HttpContextExtensions
     public static string GetAuthorizedServiceRedirectUri(this HttpContext httpContext)
         => $"{httpContext.Request.Scheme}://{httpContext.Request.Host}/umbraco/api/{nameof(AuthorizedServiceResponseController).Replace("Controller", string.Empty)}/{nameof(AuthorizedServiceResponseController.HandleIdentityResponse)}";
+
+    public static string GetOAuth1AuthorizedServiceRedirectUri(this HttpContext httpContext)
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Helpers/OAuth1aHelper.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace Umbraco.AuthorizedServices.Helpers;
+
+internal static class OAuth1aHelper
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Services/IAuthorizationRequestSender.cs:
```diff
@@ -22,4 +22,12 @@ public interface IAuthorizationRequestSender
+    /// <param name="serviceDetail">The service detail.</param>
+    /// <param name="parameters">The authorization parameters (OAuth token and verifier code).</param>
+    /// <returns></returns>
+    Task<HttpResponseMessage> SendOAuth1aRequest(ServiceDetail serviceDetail, Dictionary<string, string> parameters);
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Services/IAuthorizedRequestBuilder.cs:
```diff
@@ -43,4 +43,38 @@ public interface IAuthorizedRequestBuilder
+    /// <param name="httpMethod">The HTTP method.</param>
+    /// <param name="requestContent">The request data.</param>
+    /// <returns>The request instance.</returns>
+    HttpRequestMessage CreateIdentityRequestMessage<TRequest>(
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Services/ITokenFactory.cs:
```diff
@@ -15,4 +15,12 @@ public interface ITokenFactory
+    /// <param name="responseContent">The response content.</param>
+    /// <param name="serviceDetail">The service detail.</param>
+    /// <returns>The <see cref="OAuth1aToken"/> instance.</returns>
+    OAuth1aToken CreateFromOAuth1aResponseContent(string responseContent, ServiceDetail serviceDetail);
```
Updated

acoumb commented on src/Umbraco.AuthorizedServices/Services/Implement/AuthorizedServiceCaller.cs:
```diff
@@ -100,6 +104,20 @@ public async Task<string> SendRequestRawAsync<TRequest>(string serviceAlias, str
+            OAuth1aToken? token = GetOAuth1aToken(serviceAlias);
+            if (token is null)
+            {
+                // No token exists, query the service for a request_token.
```
The main reason behind this is to avoid making calls for a `request_token` to the service each time the service details are retrieved in case the authorization process completed successfully already and an `access_token` and `token_secret` exist.

acoumb commented on src/Umbraco.AuthorizedServices/Services/ITokenStorage.cs:
```diff
@@ -3,23 +3,24 @@
+/// Defines operations for storing <see cref="Token"/> or <see cref="OAuth1aToken"/> instances.
 /// </summary>
-public interface ITokenStorage
+public interface ITokenStorage<T>
```
Updated

acoumb commented:
With the latest commit I have also included the feature to add manually an `access_token` and a `token_secret` for `OAuth1` based services. 
For consistency, I have aligned the UI for the 3 components used to manually add details.

#### OAuth1
![image](https://github.com/umbraco/Umbraco.AuthorizedServices/assets/95346674/abb78546-6cd4-4a43-a6a4-098e7cf25acf)

#### OAuth2
![image](https://github.com/umbraco/Umbraco.AuthorizedServices/assets/95346674/11662ba7-c480-464b-a2b1-367a9534a155)

#### Api Key
![image](https://github.com/umbraco/Umbraco.AuthorizedServices/assets/95346674/a5684f85-a748-4908-a5bc-c6561bc6dc7d)

AndyButland reviewed (commented):
This is looking good, but unfortunately I wasn't able to get it to work with Flickr - I'm getting the following message back when going through the authorization process:

```
oauth_problem=parameter_absent&oauth_parameters_absent=oauth_consumer_key%26oauth_signature%26oauth_signature_method%26oauth_nonce%26oauth_timestamp
```

Which looks like we are missing sending some parameters?  Maybe you can try it and see if you get the same, and see what's wrong here.

We seem to be adding these when we are making an authorized call to the API - i.e. in `AuthorizedRequestBuilder` - but not when we are going through the authorization steps (in `AuthorizedServiceAuthorizer`).

---

I made some updates in PR #31 - a few things related to method names, fixed up some issues I was having with the migrations, added a rename table migration so we are aligned in the database as well and removed another cache implementation we had that could be replaced with using Umbraco's runtime cache.

---

I'd suggest first verifying your existing implementation to make sure the error I'm seeing with Flick is something you see too.  If not, I may have broken something!  But I don't think so as I couldn't get it to work even when I reverted back to before I made any amends.

Assuming it's not working for you, please then have a look over my updates, and if they make sense to you and you can't see any issues, merge them in here, and then see if you can see what's missing with the Flickr test case for OAuth1.

Then I'd probably like one more look over - I think we can do a bit more clean-up/consistency around creating a "parameters builder" for the OAuth1 parameters, but it'll be easier/safer to do that once there's a working sample to test against.

acoumb commented:
Additional updates included:
- when issuing sample requests, I changed to pick the path from the service detail instead of the query string, as in the case of _Flickr_ (`"/rest?nojsoncallback=1&format=json&method=flickr.test.login`) an URL contains already query string parameters that got removed.
- added a flag called `UseRequestTokenWithExtendedParametersList` to toggle the append operation of additional parameters when requesting the access token. And it will take into consideration the `RequestTokenMethod` - usually the services use a `POST` request, _Flicker_ uses a `GET`.
- on receiving the `oauth_token_secret`, _Flickr_ uses it for retrieving the access token, therefor I cache it with the `oauth_token`.

AndyButland reviewed (approved):
Great, I've verified it's all working now and the amends look good.

The only thing of significance I changed was to remove the `UseRequestTokenWithExtendedParametersList` setting you had added.  I tested just sending all parameters to both Flickr and Twitter, and both worked, so seemed to me we could just to do that.  If some aren't needed for some services, that doesn't really matter, as long as they are just ignored, which they seem to be.  And then it's one less configuration option for people to understand whether to use or not.
